File: Difficulty: Medium/Coin Change (Minimum Coins)/coin-change-minimum-coins.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    int minCoins(vector<int> &coins, int sum) {
        // code here
        // Create a DP array to store the minimum coins
        // for each value up to sum
        vector<int> dp(sum+1, INT_MAX);
        
        // Base case: 0 coins are needed to make sum 0
        dp[0] = 0; 
        
        // Iterate over each coin in the coins array
        for (int i=coins.size()-1; i>=0; i--) {
            
            // Iterate through all sums from 1 to the given sum
            for (int j=1; j<=sum; j++) {
                
                // Initialize take and noTake variables
                int take = INT_MAX, noTake = INT_MAX;
                
                // Check if we can take the current coin
                if (j-coins[i]>=0 && coins[i]>0) {
                    
                    // Get the minimum coins needed for
                    // the remaining sum
                    take = dp[j-coins[i]]; 
                    
                    // Increment take if it's a valid option
                    if (take != INT_MAX) take++; 
                }
                
                // If we have more coins available,
                // consider not taking the current coin
                
                if (i+1<coins.size())
                // Get the minimum coins needed without
                // taking the current coi
                    noTake = dp[j]; 
                
                // Update the DP array with the minimum of
                // taking or not taking the coin
                dp[j] = min(take, noTake);
            }
        }
        
        // found a solution, return the minimum
        // coins needed, otherwise return -1
        if (dp[sum]!=INT_MAX) return dp[sum];
        return -1;
    }
};


//{ Driver Code Starts.

int main() {
    string ts;
    getline(cin, ts);
    int t = stoi(ts);
    while (t--) {
        string ks;
        getline(cin, ks);
        int k = stoi(ks);
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        int res = obj.minCoins(arr, k);
        cout << res << endl;
        cout << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Coin Change (Minimum Coins)/coin-change-minimum-coins.java
//{ Driver Code Starts
// Initial Template for Java
import java.io.*;
import java.lang.*;
import java.util.*;


// } Driver Code Ends

class Solution {

    public int minCoins(int coins[], int sum) {
        // code here
        int n=coins.length;
        int [][] dp=new int [n][sum+1];
        for(int i=0;i<dp.length;i++)
        {
            for(int j=0;j<dp[0].length;j++)
            {
                dp[i][j]=-1;
            }
        }
        int ans=solve(dp,coins,sum,0,0);
        return ans==Integer.MAX_VALUE?-1:ans;
    }
    public int solve(int [][]dp, int [] coins, int sum, int curr,int i) 
    {
    if (sum == curr) return 0;
    if (i >= coins.length || sum < curr) return Integer.MAX_VALUE;
    if (dp[i][curr] != -1) return dp[i][curr];

    int take = Integer.MAX_VALUE;
    if (sum >= curr + coins[i]) {
        int res = solve(dp, coins, sum, curr + coins[i], i);
        if (res != Integer.MAX_VALUE) take = 1 + res;
    }

    int skip = solve(dp, coins, sum, curr, i + 1);

    return dp[i][curr] = Math.min(skip, take);
}
}


//{ Driver Code Starts.

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            int k = Integer.parseInt(br.readLine());
            String line = br.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;
            Solution obj = new Solution();
            int res = obj.minCoins(arr, k);

            System.out.println(res);

            System.out.println("~");
        }
    }
}

// } Driver Code Ends